external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala

```diff
@@ -135,23 +135,79 @@ private[kafka010] class KafkaOffsetReader(
-    val fetched = runUninterruptibly {
-      withRetriesWithoutInterrupt {
-        // Poll to get the latest assigned partitions
-        consumer.poll(0)
-        val partitions = consumer.assignment()
+    val fnAssertParametersWithPartitions: ju.Set[TopicPartition] => Unit = { partitions =>
+      assert(partitions.asScala == partitionOffsets.keySet,
+        "If startingOffsets contains specific offsets, you must specify all TopicPartitions.\n" +
+          "Use -1 for latest, -2 for earliest, if you don't care.\n" +
+          s"Specified: ${partitionOffsets.keySet} Assigned: ${partitions.asScala}")
+      logDebug(s"Partitions assigned to consumer: $partitions. Seeking to $partitionOffsets")
+    }
 
-        // Call `position` to wait until the potential offset request triggered by `poll(0)` is
-        // done. This is a workaround for KAFKA-7703, which an async `seekToBeginning` triggered by
-        // `poll(0)` may reset offsets that should have been set by another request.
-        partitions.asScala.map(p => p -> consumer.position(p)).foreach(_ => {})
+    val fnRetrievePartitionOffsets: ju.Set[TopicPartition] => Map[TopicPartition, Long] = { _ =>
+      partitionOffsets
+    }
+
+    val fnAssertFetchedOffsets: Map[TopicPartition, Long] => Unit = { fetched =>
+      partitionOffsets.foreach {
+        case (tp, off) if off != KafkaOffsetRangeLimit.LATEST &&
+          off != KafkaOffsetRangeLimit.EARLIEST =>
+          if (fetched(tp) != off) {
+            reportDataLoss(
+              s"startingOffsets for $tp was $off but consumer reset to ${fetched(tp)}")
+          }
+        case _ =>
+        // no real way to check that beginning or end is reasonable
+      }
+    }
+
+    fetchSpecificOffsets0(fnAssertParametersWithPartitions, fnRetrievePartitionOffsets,
+      fnAssertFetchedOffsets)
+  }
+
+  def fetchSpecificTimestampBasedOffsets(topicTimestamps: Map[String, Long]): KafkaSourceOffset = {
+    val fnAssertParametersWithPartitions: ju.Set[TopicPartition] => Unit = { partitions =>
+      val assignedTopics = partitions.asScala.map(_.topic())
+      assert(assignedTopics == topicTimestamps.keySet,
+        "If starting/endingOffsetsByTimestamp contains specific offsets, you must specify all " +
+          s"topics. Specified: ${topicTimestamps.keySet} Assigned: $assignedTopics")
+      logDebug(s"Partitions assigned to consumer: $partitions. Seeking to $topicTimestamps")
+    }
+
+    val fnRetrievePartitionOffsets: ju.Set[TopicPartition] => Map[TopicPartition, Long] = {
+      partitions => {
+        val partitionTimestamps: ju.Map[TopicPartition, java.lang.Long] =
+          partitions.asScala.map { topicAndPartition =>
+            topicAndPartition -> java.lang.Long.valueOf(topicTimestamps(topicAndPartition.topic()))
+          }.toMap.asJava
+
+        val offsetForTime: ju.Map[TopicPartition, OffsetAndTimestamp] =
+          consumer.offsetsForTimes(partitionTimestamps)
```

Comment:
> If the message format version in a partition is before 0.10.0, i.e. the messages do not have timestamps, null will be returned for that partition. You mentioned it's not probable that the broker is older than 0.10.0 which I tend to agree but here the broker is new but the handled Kafka log is old.

We cannot distinguish whether there's no offset which satisfies the timestamp condition or the log is written as old message format. For the first case I think `LATEST` should be provided, and due to lack of information we have to apply this to second case as well.

> This suffers from a classic Kafka issue, namely this method may block indefinitely if the partition does not exist.

I'm not an expert of Kafka so hard to say, but we run `consumer.poll` and `consumer.assignment` just before calling this and assert on partitions, so I feel the chance of non-exist partition would be pretty rare.

Reply:
> For the first case I think LATEST should be provided, and due to lack of information we have to apply this to second case as well.

I've taken a deeper look at Kafka consumer and it works as we've discussed. It's fine.

> we run consumer.poll and consumer.assignment

The actual implementation is good as long as the user is not able to provide TopicPartitions without some control. I've added this here because of my [comment](https://github.com/apache/spark/pull/23747#discussion_r266412216). If we agree on having topic only and not TopicPartition as parameter then this discussion could be resolved.

Reply:
Default-to-LATEST makes some sense for ending offsets, but does it make sense for starting offsets? It's weird that the stream might return records with timestamps below the specified start point.

Reply:
This is actually due to the behavior of `offsetsForTimes`, but I agree your point. Would we want to let query crash if there's no matched record for starting timestamp offsets?

Reply:
I think that's better than latest. I'd also be fine with just delaying for some timeout to see if the timestamp shows up (I imagine that's what most people who directly call offsetsForTimes would do upon receiving null), but that gets complicated enough that it probably wouldn't make sense as part of the PR.

Reply:
Right now I'm now seeing other view, like data skew. Some of partitions may not have any records while other partitions have, or at least maybe far behind. At that case we may not want to crash the query but also don't want to drag the query being run.

Most of the cases, end users would put timestamp as backwards, which you may not expect to get no records matching one but can get later. (I guess it could happen based on the meaning of Kafka timestamp - I meant `mode`, but they will mostly roll back hours or even days which there's rare to encounter the case.)

Reply:
I can definitely see why the default-to-latest behavior could be helpful for things like data skew. But when push comes to shove, if I specify startingTimestamp = 200 and the output contains timestamp 195, I think users are going to consider that a pretty straightforward bug. The underlying Kafka consumer library doesn't have this behavior; it will keep returning null forever until an offset after the starting timestamp becomes available.

Reply:
As we see both possibilities: maybe we could have an option which makes this case failing the query by default with proposing log message to toggle if they really want to just start reading. I'll address it soon.

Reply:
Think about this more: need to check whether it's possible to do, but the ideal option is not reading from such partitions. We may have thought we should read from all partitions we've assigned, so have to choose between failing the query or taking the risk to read latest. Ideally we shouldn't do that.

The problem is, for streaming query, the option is only used for once and ignored afterwards, which will break the intention what we are trying to do. For timestamp offset case, it should still try to pull the offsets based on timestamp for every batch, and compare it with offsets in checkpoint if exists to see whether we can read from such partition or not.

This sounds me as a non-trivial effort, so maybe we need to start with just failing the query for taking safest approach, and address above in follow-up issue.

@jose-torres What do you think? cc. @koeninger @gaborgsomogyi

Reply:
Just addressed the first phase: fail the query. Other phase we can try with follow-up issue.